app/models/manageiq/providers/oracle/infra_manager/refresh_parser.rb

```diff
@@ -0,0 +1,505 @@
+      switches, switch_uids[mor], lan_uids[mor] = server_inv_to_switch_hashes(server_inv)
+
+      hardware = server_inv_to_hardware_hash(server_inv)
+      hardware[:guest_devices], guest_device_uids[mor] = server_inv_to_guest_device_hashes(server_inv, switch_uids[mor])
+      hardware[:networks] = server_inv_to_network_hashes(server_inv, guest_device_uids[mor])
+
+      new_result = {
+        :type             => 'ManageIQ::Providers::Oracle::InfraManager::Host',
+        :ems_ref          => uri_to_ref(server_inv.uri),
+        :ems_ref_obj      => uri_to_ref(server_inv.uri),
+        :name             => server_inv.name || hostname,
+        :hostname         => hostname,
+        :ipaddress        => ipaddress,
+        :uid_ems          => server_inv.key,
+        :vmm_vendor       => 'oracle',
+        :vmm_product      => 'oraclevm',
+        :vmm_version      => server_inv.ovm_version.split('-').first,
+        :vmm_buildnumber  => server_inv.ovm_version.split('-').last,
+        :connection_state => connection_state,
+        :power_state      => power_state,
+
+        :operating_system => server_inv_to_os_hash(server_inv, hostname),
+
+        :ems_cluster      => cluster_uids[server_inv.server_pool_id["value"] + "_cluster"],
+        :hardware         => hardware,
+        :switches         => switches,
+
+      }
+
+      result << new_result
+      result_uids[mor] = new_result
+    end
+
+    return result, result_uids, lan_uids, switch_uids, guest_device_uids, scsi_lun_uids
+  end
+
+  def self.server_inv_to_hardware_hash(inv)
+    return nil if inv.nil?
+
+    cpus = inv.cpus
+
+    return nil if cpus.empty?
+
+    cpu = cpus[0]
+
+    {
+      :cpu_type             => cpu.model_name,
+      :manufacturer         => cpu.vendor_id,
+      :model                => cpu.model_name,
+      :cpu_speed            => 0,
+      :memory_mb            => inv.memory,
+      :cpu_sockets          => 1,
+      :cpu_total_cores      => cpus.length,
+      :cpu_cores_per_socket => cpus.length
+    }
+  end
+
+  def self.server_inv_to_switch_hashes(inv)
+    ports = inv.ethernet_ports
+
+    result = []
+    result_uids = {:pnic_id => {}}
+    lan_uids    = {}
+    return result, result_uids if ports.nil?
+
+    ports.each do |port|
+      network = port.network
+
+      next if network.nil?
+
+      network_id = network.key
+
+      lan = {:name => network.name, :uid_ems => network_id, :tag => nil}
+      lan_uids[network_id] = lan
+      new_result = {
+        :uid_ems => network_id,
+        :name    => network.name,
+        :lans    => [lan]
+      }
+
+      result << new_result
+      result_uids[network_id] = new_result
+    end
+
+    return result, result_uids, lan_uids
+  end
+
+  def self.server_inv_to_guest_device_hashes(inv, switch_uids)
+    ports = inv.ethernet_ports
+
+    result = []
+    result_uids = {}
+    return result, result_uids if ports.nil?
+
+    result_uids[:pnic] = {}
+    ports.each do |port|
```

Comment:
As a general rule of thumb, try and use `map` over `each`. Especially when the purpose is to output something you are converting. As, `each_with_object()` is useful when you are building a hash from a list of items

```ruby
result = ports.map do |port|
  network = port.network
  switch = switch_uids[network.key] if network
  {
    :uid_ems         => port.key,
    :device_name     => port.name,
    :device_type     => 'ethernet',
    :location        => port.interface_name,
    :present         => true,
    :controller_type => 'ethernet',
  }.tap do |new_result|
    new_result[:switch] = switch if switch
    result_uids[:pnic][port.key] = new_result
  end
end
```

If you feel the `tap` is too complex, it could look like this:

```ruby
result = ports.map do |port|
  network = port.network
  switch = switch_uids[network.key] if network
  new_result = {
    :uid_ems         => port.key,
    :device_name     => port.name,
    :device_type     => 'ethernet',
    :location        => port.interface_name,
    :present         => true,
    :controller_type => 'ethernet',
  }
  new_result[:switch] = switch if switch
  result_uids[:pnic][port.key] = new_result
  new_result
end
```